dali/pipeline/util/operator_impl_utils.h

```diff
@@ -51,6 +55,123 @@ class OpImplBase<CPUBackend> {
   }
 };
 
+
+namespace detail {
+
+template <typename T>
+void ReadArgConstant(std::vector<T> &out, const string &arg_name, const OpSpec &spec) {
+  out.resize(1);
+  out[0] = spec.GetArgument<T>(arg_name);
+}
+
+template <typename T>
+void ReadArgInput(std::vector<T> &out, const std::string &arg_name,
+                  const OpSpec &spec, const workspace_t<CPUBackend> &ws) {
+  const auto& arg_in = ws.ArgumentInput(arg_name);
+  auto arg_in_view = view<const float>(arg_in);
+  DALI_ENFORCE(is_uniform(arg_in_view.shape) && volume(arg_in_view.shape[0]) == 1,
+    make_string("``", arg_name, "`` must be a scalar or a 1D tensor with a single element"));
+  if (arg_in_view.shape.sample_dim() > 0) {
+    DALI_WARN_ONCE("Warning: \"", arg_name, "\""
+                   " expected a scalar but received a 1D tensor with a single "
+                   "element. Please use a scalar instead.");
+  }
+
+  auto nsamples = arg_in_view.size();
+  out.resize(nsamples);
+  for (int i = 0; i < nsamples; i++) {
+    out[i] = arg_in_view[i].data[0];
+  }
+}
+
+template <typename T>
+void ReadArgInput(std::vector<std::vector<T>> &out, const std::string &arg_name,
+                  const OpSpec &spec, const workspace_t<CPUBackend> &ws) {
+  const auto& arg_in = ws.ArgumentInput(arg_name);
+  auto arg_in_view = view<const float>(arg_in);
+  DALI_ENFORCE(is_uniform(arg_in_view.shape),
+    make_string("All samples in argument ``", arg_name, "`` should have the same shape"));
+  DALI_ENFORCE(arg_in_view.shape.sample_dim() == 1,
+    make_string("``", arg_name, "`` must be a 1D tensor"));
+
+  auto nsamples = arg_in_view.size();
+  out.resize(nsamples);
+  for (int i = 0; i < nsamples; i++) {
+    auto ndim = arg_in_view[i].shape[0];
+    out[i].resize(ndim);
+    for (int d = 0; d < ndim; d++) {
+      out[i][d] = arg_in_view[i].data[d];
+    }
+  }
+}
+
+}  // namespace detail
+
+
+template <typename ArgType>
+class ArgHelper {
+ public:
+  ArgHelper(const std::string &arg_name, const OpSpec &spec)
+      : arg_name_(arg_name),
```

Comment:
```suggestion
  ArgHelper(std::string arg_name, const OpSpec &spec)
      : arg_name_(std::move(arg_name)),
```
Paradoxically, passing the string by value in this case can _save_ a copy. Consider
```C++
ArgHelper("literal", spec);```
^^^ this creates a string (by copying the literal) and then copies it again to the member.
When passing by value, you create a string (by copy) and then move it to the member.
This trick is only worthwhile when the value is moved (which is the case here).